src/wayland/data_device/mod.rs

```diff
@@ -0,0 +1,418 @@
+                    if dd.client().map(|c| !c.equals(client)).unwrap_or(true) {
+                        continue;
+                    }
+                    let log = self.log.clone();
+                    let offer_meta = meta.clone();
+                    let callback = dd.user_data::<DataDeviceData>().unwrap().callback.clone();
+                    // create a corresponding data offer
+                    let offer = client
+                        .create_resource::<wl_data_offer::WlDataOffer>(dd.version())
+                        .unwrap()
+                        .implement(
+                            move |req, _offer| match req {
+                                wl_data_offer::Request::Receive { fd, mime_type } => {
+                                    // check if the associated mime type is valid
+                                    if !offer_meta.mime_types.contains(&mime_type) {
+                                        // deny the receive
+                                        debug!(log, "Denying a wl_data_offer.receive with invalid source.");
+                                        let _ = ::nix::unistd::close(fd);
+                                    } else {
+                                        (&mut *callback.lock().unwrap())(DataDeviceEvent::SendSelection {
+                                            mime_type,
+                                            fd,
+                                        });
+                                    }
+                                }
+                                _ => { /* seleciton data offers only care about the `receive` event */ }
+                            },
+                            None::<fn(_)>,
+                            (),
+                        );
+                    // advertize the offer to the client
+                    dd.send(wl_data_device::Event::DataOffer { id: offer.clone() });
+                    for mime_type in meta.mime_types.iter().cloned() {
+                        offer.send(wl_data_offer::Event::Offer { mime_type })
+                    }
+                    dd.send(wl_data_device::Event::Selection { id: Some(offer) });
+                }
+            }
+        }
+    }
+}
+
+impl SeatData {
+    fn new(log: ::slog::Logger) -> SeatData {
+        SeatData {
+            known_devices: Vec::new(),
+            selection: Selection::Empty,
+            log,
+            current_focus: None,
+        }
+    }
+}
+
+/// Initialize the data device global
+///
+/// You need to provide a `(DndAction, DndAction) -> DndAction` closure that will arbitrate
+/// the choice of action resulting from a drag'n'drop session. Its first argument is the set of
+/// available actions (which is the intersection of the actions supported by the source and targets)
+/// and the second argument is the preferred action reported by the target. If no action should be
+/// chosen (and thus the drag'n'drop should abort on drop), return `DndAction::empty()`.
```

Comment:
The callback argument does not seem to be explained?
And in the anvil implementation it is empty.

I assume this is this callback:
`- You can provide a callback closure to `init_data_device` to peek into the the actions of your clients`
from the module description?

Is this also where I would handle a server drag'n'drop?
Could maybe document how that would work exactly somewhere?

Reply:
That's what you get for writing docs past midnight I guess. I'll detail all this more.